# Patch to enable shared library building
# Copyright (C) 2010-2015 Georgi D. Sotirov <gdsotirov@dir.bg>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
#
diff -urNad lua-5.1.5-orig/Makefile lua-5.1.5/Makefile
--- lua-5.1.5-orig/Makefile	2012-02-10 11:50:23.000000000 +0200
+++ lua-5.1.5/Makefile	2015-05-02 12:15:28.000000000 +0300
@@ -43,7 +43,7 @@
 # What to install.
 TO_BIN= lua luac
 TO_INC= lua.h luaconf.h lualib.h lauxlib.h ../etc/lua.hpp
-TO_LIB= liblua.a
+TO_LIB= liblua.a liblua.so
 TO_MAN= lua.1 luac.1
 
 # Lua version and release.
diff -urNad lua-5.1.5-orig/src/Makefile lua-5.1.5/src/Makefile
--- lua-5.1.5-orig/src/Makefile	2012-02-13 22:41:22.000000000 +0200
+++ lua-5.1.5/src/Makefile	2015-05-02 12:16:08.000000000 +0300
@@ -23,11 +23,17 @@
 PLATS= aix ansi bsd freebsd generic linux macosx mingw posix solaris
 
 LUA_A=	liblua.a
+LUA_SO= liblua.so
 CORE_O=	lapi.o lcode.o ldebug.o ldo.o ldump.o lfunc.o lgc.o llex.o lmem.o \
 	lobject.o lopcodes.o lparser.o lstate.o lstring.o ltable.o ltm.o  \
 	lundump.o lvm.o lzio.o
+CORE_PO=lapi.po lcode.po ldebug.po ldo.po ldump.po lfunc.po lgc.po llex.po lmem.po \
+	lobject.po lopcodes.po lparser.po lstate.po lstring.po ltable.po ltm.po  \
+	lundump.po lvm.po lzio.po
 LIB_O=	lauxlib.o lbaselib.o ldblib.o liolib.o lmathlib.o loslib.o ltablib.o \
 	lstrlib.o loadlib.o linit.o
+LIB_PO=	lauxlib.po lbaselib.po ldblib.po liolib.po lmathlib.po loslib.po ltablib.po \
+	lstrlib.po loadlib.po linit.po
 
 LUA_T=	lua
 LUA_O=	lua.o
@@ -35,8 +41,8 @@
 LUAC_T=	luac
 LUAC_O=	luac.o print.o
 
-ALL_O= $(CORE_O) $(LIB_O) $(LUA_O) $(LUAC_O)
-ALL_T= $(LUA_A) $(LUA_T) $(LUAC_T)
+ALL_O= $(CORE_O) $(CORE_PO) $(LIB_O) $(LIB_PO) $(LUA_O) $(LUAC_O)
+ALL_T= ${LUA_A} ${LUA_SO} ${LUA_T} ${LUAC_T}
 ALL_A= $(LUA_A)
 
 default: $(PLAT)
@@ -47,16 +53,25 @@
 
 a:	$(ALL_A)
 
+%.o : %.c
+	$(CC) $(CFLAGS) -o $@ -c $<
+
+%.po : %.c
+	$(CC) $(CFLAGS) -fPIC -o $@ -c $<
+
 $(LUA_A): $(CORE_O) $(LIB_O)
 	$(AR) $@ $(CORE_O) $(LIB_O)	# DLL needs all object files
 	$(RANLIB) $@
 
-$(LUA_T): $(LUA_O) $(LUA_A)
+$(LUA_T): $(LUA_O) $(LUA_SO)
 	$(CC) -o $@ $(MYLDFLAGS) $(LUA_O) $(LUA_A) $(LIBS)
 
-$(LUAC_T): $(LUAC_O) $(LUA_A)
+$(LUAC_T): $(LUAC_O) $(LUA_SO) $(LUA_A)
 	$(CC) -o $@ $(MYLDFLAGS) $(LUAC_O) $(LUA_A) $(LIBS)
 
+$(LUA_SO): $(CORE_PO) $(LIB_PO)
+	$(CC) -o $@ -shared -lm -ldl $?
+
 clean:
 	$(RM) $(ALL_T) $(ALL_O)
 
@@ -96,7 +111,7 @@
 	$(MAKE) all MYCFLAGS=
 
 linux:
-	$(MAKE) all MYCFLAGS=-DLUA_USE_LINUX MYLIBS="-Wl,-E -ldl -lreadline -lhistory -lncurses"
+	$(MAKE) all MYCFLAGS="$(MYCFLAGS)" MYLIBS="-Wl,-E -ldl -lreadline -lhistory -lncurses"
 
 macosx:
 	$(MAKE) all MYCFLAGS=-DLUA_USE_LINUX MYLIBS="-lreadline"
